File: SortVision/public/code/bubble/java/bubbleSort.java
/**
 * A utility class to implement the Bubble Sort algorithm in Java.
 * 
 * <p>This class provides a method to sort an integer array in ascending order
 * using the Bubble Sort algorithm with early-exit optimization.</p>
 * 
 * Time Complexity:
 * - Worst case: O(n^2)
 * - Best case (optimized with no swaps): O(n)
 * 
 * Space Complexity: O(1) (in-place sorting)
 */
public class BubbleSort {

    /**
     * Sorts the input array in ascending order using the Bubble Sort algorithm.
     *
     * @param arr the array to be sorted
     * @throws IllegalArgumentException if the input array is null
     */
    public static void bubbleSort(int[] arr) {
        if (arr == null) {
            throw new IllegalArgumentException("Input array cannot be null.");
        }

        int n = arr.length;

        // No need to sort if array has 0 or 1 elements
        if (n <= 1) {
            return;
        }

        // Outer loop for each pass
        for (int i = 0; i < n - 1; i++) {
            boolean swapped = false;

            // Inner loop for comparing adjacent elements
            for (int j = 0; j < n - 1 - i; j++) {
                if (arr[j] > arr[j + 1]) {
                    // Swap elements if they are in the wrong order
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                    swapped = true;
                }
            }

            // Optimization: If no swaps were made, array is already sorted
            if (!swapped) {
                break;
            }
        }
    }

    /**
     * Prints the contents of an array.
     * 
     * @param arr the array to print
     */
    public static void printArray(int[] arr) {
        if (arr == null || arr.length == 0) {
            System.out.println("[]");
            return;
        }

        System.out.print("[");
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i]);
            if (i < arr.length - 1) {
                System.out.print(", ");
            }
        }
        System.out.println("]");
    }

    /**
     * Demonstrates example usage and test cases for the BubbleSort class.
     *
     * Includes:
     * - Empty array
     * - Single element
     * - Already sorted array
     * - Reverse sorted array
     * - Random unordered array
     */
    public static void main(String[] args) {
        // Test case 1: Empty array
        int[] test1 = {};
        bubbleSort(test1);
        System.out.print("Test 1 (Empty): ");
        printArray(test1); // []

        // Test case 2: Single element
        int[] test2 = {42};
        bubbleSort(test2);
        System.out.print("Test 2 (Single): ");
        printArray(test2); // [42]

        // Test case 3: Already sorted
        int[] test3 = {1, 2, 3, 4, 5};
        bubbleSort(test3);
        System.out.print("Test 3 (Sorted): ");
        printArray(test3); // [1, 2, 3, 4, 5]

        // Test case 4: Reverse sorted
        int[] test4 = {9, 7, 5, 3, 1};
        bubbleSort(test4);
        System.out.print("Test 4 (Reverse): ");
        printArray(test4); // [1, 3, 5, 7, 9]

        // Test case 5: Random order
        int[] test5 = {5, 2, 8, 3, 1};
        bubbleSort(test5);
        System.out.print("Test 5 (Random): ");
        printArray(test5); // [1, 2, 3, 5, 8]
    }
}

File: SortVision/public/code/bubble/c/bubbleSort.c
#include <stdio.h>
#include <stdbool.h>

/**
 * Bubble Sort implementation
 * 
 * Sorts the input array in ascending order using the bubble sort algorithm.
 * Optimized with early termination if the array becomes sorted early.
 * 
 * Time Complexity:
 * - Best: O(n)       [already sorted]
 * - Average: O(n^2)
 * - Worst: O(n^2)
 * 
 * Space Complexity: O(1) [in-place]
 * 
 * @param arr The array to sort
 * @param n   The number of elements in the array
 */
void bubbleSort(int arr[], int n) {
    if (n <= 1) return;  // Handle empty or single-element arrays

    for (int i = 0; i < n - 1; i++) {
        bool swapped = false;

        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                // Swap elements
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;

                swapped = true;
            }
        }

        // If no swaps happened, array is already sorted
        if (!swapped) break;
    }
}

/**
 * Utility function to print an array
 */
void printArray(int arr[], int n) {
    printf("[");
    for (int i = 0; i < n; i++) {
        printf("%d", arr[i]);
        if (i != n - 1) printf(", ");
    }
    printf("]\n");
}

/**
 * Example test cases for Bubble Sort
 */
void runTests() {
    int test1[0];
    int test2[] = {42};
    int test3[] = {5, 2, 9, 1, 5};
    int test4[] = {1, 2, 3, 4, 5};
    int test5[] = {10, -5, 3, 0, 2};

    int* tests[] = {test1, test2, test3, test4, test5};
    int sizes[] = {0, 1, 5, 5, 5};
    const char* descriptions[] = {
        "Empty array",
        "Single element",
        "Unsorted with duplicates",
        "Already sorted",
        "Contains negatives"
    };

    for (int i = 0; i < sizeof(tests)/sizeof(tests[0]); i++) {
        printf("Test %d (%s):\nBefore: ", i + 1, descriptions[i]);
        printArray(tests[i], sizes[i]);

        bubbleSort(tests[i], sizes[i]);

        printf("After:  ");
        printArray(tests[i], sizes[i]);
        printf("-------------------------\n");
    }
}

/**
 * Main function to demonstrate usage
 */
int main() {
    runTests();
    return 0;
}
